src/vaadin-grid-column.html

```diff
@@ -470,9 +470,25 @@
         this.parentElement._columnPropChanged('hidden', hidden);
       }
 
-      this._allCells.forEach(cell => this._toggleAttribute('hidden', hidden, cell));
-
       if (!!hidden !== !!this._previousHidden && this._grid) {
+        if (hidden === true) {
+          this._allCells.forEach(cell => {
+            if (cell._content.parentNode) {
+              cell._content.parentNode.removeChild(cell._content);
+              cell._vacant = true;
```

Comment:
The optimization seems to work well. These lines still need tests though.

Reply:
Removed setting cell vacant as it is being done at `_rowUpdate` method already. Added test to check children count before and after a column is hidden.